pkg/manifest/config/validate.go

```diff
@@ -0,0 +1,188 @@
+package config
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mkmik/multierror"
+	"k8s.io/apimachinery/pkg/api/validation"
+	"knative.dev/pkg/apis"
+)
+
+type validateVisitor struct {
+	errs         []error
+	envNames     map[string]bool
+	appNames     map[string]bool
+	serviceNames map[string]bool
+}
+
+func (m *Manifest) Validate() error {
+	vv := &validateVisitor{errs: []error{}, envNames: map[string]bool{}, appNames: map[string]bool{}, serviceNames: map[string]bool{}}
+
+	m.Walk(vv)
+	return multierror.Join(vv.errs)
+}
+
+func (vv *validateVisitor) Environment(env *Environment) error {
+	envPath := yamlPath(PathForEnvironment(env))
+	if err := checkDuplicate(env.Name, envPath, vv.envNames); err != nil {
+		vv.errs = append(vv.errs, err)
+	}
+	if err := validateName(env.Name, envPath); err != nil {
+		vv.errs = append(vv.errs, err)
+	}
+	if err := validatePipelines(env.Pipelines, envPath); err != nil {
+		vv.errs = append(vv.errs, err...)
+	}
+	return nil
+}
+
+func (vv *validateVisitor) Application(env *Environment, app *Application) error {
+	appPath := yamlPath(PathForApplication(env, app))
+	if err := checkDuplicate(app.Name, appPath, vv.appNames); err != nil {
+		vv.errs = append(vv.errs, err)
+	}
+
+	if err := validateName(app.Name, appPath); err != nil {
+		vv.errs = append(vv.errs, err)
+	}
+
+	if app.Services == nil && app.ConfigRepo == nil {
+		vv.errs = append(vv.errs, missingFieldsError([]string{"services", "config_repo"}, []string{appPath}))
+	}
+	if app.Services != nil && app.ConfigRepo != nil {
+		vv.errs = append(vv.errs, apis.ErrMultipleOneOf(yamlJoin(appPath, "services"), yamlJoin(appPath, "config_repo")))
+	}
+	if app.ConfigRepo != nil {
+		vv.errs = append(vv.errs, validateConfigRepo(app.ConfigRepo, yamlJoin(appPath, "config_repo"))...)
+	}
```

Comment:
If application has configRepo, validation fails (regardless services exist or not)?